[WIP] Perf-testing on hash functions

wingo opened the pull request:
This is a WIP PR not intended to be merged.  It cherry-picks b00fbd8d8d9a572b3791dbf64127b79ec6981bbf from `lwaftr` to add the `snabbmark hash` function, extends the hash test to test variable-length keys, then adds the fnv hash.

Results:
```
$ for i in 1 2 4 6 8 12 16 24 32; do echo "./snabb snabbmark hash $i"; sudo ./snabb snabbmark hash $i; echo; done
./snabb snabbmark hash 1
No PMU available: single core cpu affinity required
baseline: 2.84 ns per iteration (result: 0)
jenkins hash: 18.28 ns per iteration (result: 236129928)
fnv hash: 4.42 ns per iteration (result: 2590564124)
murmur hash (32 bit): 17.93 ns per iteration (result: 1364076727)

./snabb snabbmark hash 2
No PMU available: single core cpu affinity required
baseline: 3.49 ns per iteration (result: 0)
jenkins hash: 24.06 ns per iteration (result: 2782625322)
fnv hash: 5.40 ns per iteration (result: 790051092)
murmur hash (32 bit): 18.25 ns per iteration (result: 821347078)

./snabb snabbmark hash 4
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 4.54 ns per iteration (result: 2355983250)
fnv hash: 6.34 ns per iteration (result: 3420582196)
murmur hash (32 bit): 8.22 ns per iteration (result: 593689054)

./snabb snabbmark hash 6
No PMU available: single core cpu affinity required
baseline: 3.21 ns per iteration (result: 0)
jenkins hash: 23.69 ns per iteration (result: 3787970914)
fnv hash: 7.11 ns per iteration (result: 3875976788)
murmur hash (32 bit): 18.18 ns per iteration (result: 4065731677)

./snabb snabbmark hash 8
No PMU available: single core cpu affinity required
baseline: 3.19 ns per iteration (result: 0)
jenkins hash: 7.49 ns per iteration (result: 236129928)
fnv hash: 8.16 ns per iteration (result: 3645689972)
murmur hash (32 bit): 10.14 ns per iteration (result: 1669671676)

./snabb snabbmark hash 12
No PMU available: single core cpu affinity required
baseline: 3.18 ns per iteration (result: 0)
jenkins hash: 103.57 ns per iteration (result: 3787970914)
fnv hash: 11.70 ns per iteration (result: 1767269300)
murmur hash (32 bit): 107.42 ns per iteration (result: 3644920907)

./snabb snabbmark hash 16
No PMU available: single core cpu affinity required
baseline: 3.20 ns per iteration (result: 0)
jenkins hash: 110.06 ns per iteration (result: 1799890284)
fnv hash: 14.04 ns per iteration (result: 2754996980)
murmur hash (32 bit): 114.29 ns per iteration (result: 2167721464)

./snabb snabbmark hash 24
No PMU available: single core cpu affinity required
baseline: 3.32 ns per iteration (result: 0)
jenkins hash: 119.25 ns per iteration (result: 2548393908)
fnv hash: 21.12 ns per iteration (result: 3617052020)
murmur hash (32 bit): 117.20 ns per iteration (result: 3708130561)

./snabb snabbmark hash 32
No PMU available: single core cpu affinity required
baseline: 2.71 ns per iteration (result: 0)
jenkins hash: 132.02 ns per iteration (result: 333401968)
fnv hash: 31.12 ns per iteration (result: 666669044)
murmur hash (32 bit): 129.31 ns per iteration (result: 407949976)

```

Conclusions: Although the pure-Lua fast cases for the Jenkins hash (for 4 and 8 bytes) perform very well, the generic version does not get specialized by LuaJIT to nice code.  I am not quite sure why but it does seem that the DynASM FNV hash is doing much better where we don't hit the Jenkins fast case.  Murmur is worse in all situations (from a speed POV).

Still it would seem that we can do better than the current FNV hash, especially considering that it's not unrolled and does byte-by-byte access.  It's worthy of a bit of code golf :)

wingo committed:
Add hash and ctable subcommands to snabbmark

wingo committed:
Allow "snabbmark hash" to test different key sizes

Instead of always hashing a single uint32, "snabbmark hash" hashes a
byte vector which is filled each iteration.  This imposes some
overhead but lets us get closer to checking ctable performance.

wingo committed:
add fnv hash to "snabbmark hash"

wingo committed:
Unroll FNV hash function

wingo commented:
Added a commit to unroll the hash function.  New results:

```
$ for i in 1 2 4 6 8 12 16 24 32; do echo "./snabb snabbmark hash $i"; sudo ./snabb snabbmark hash $i; echo; done
./snabb snabbmark hash 1
No PMU available: single core cpu affinity required
baseline: 2.84 ns per iteration (result: 0)
jenkins hash: 18.23 ns per iteration (result: 236129928)
fnv hash: 4.12 ns per iteration (result: 2590564124)
murmur hash (32 bit): 17.93 ns per iteration (result: 1364076727)

./snabb snabbmark hash 2
No PMU available: single core cpu affinity required
baseline: 3.49 ns per iteration (result: 0)
jenkins hash: 24.07 ns per iteration (result: 2782625322)
fnv hash: 4.78 ns per iteration (result: 790051092)
murmur hash (32 bit): 18.23 ns per iteration (result: 821347078)

./snabb snabbmark hash 4
No PMU available: single core cpu affinity required
baseline: 3.17 ns per iteration (result: 0)
jenkins hash: 4.42 ns per iteration (result: 2355983250)
fnv hash: 4.74 ns per iteration (result: 3420582196)
murmur hash (32 bit): 8.04 ns per iteration (result: 593689054)

./snabb snabbmark hash 6
No PMU available: single core cpu affinity required
baseline: 3.17 ns per iteration (result: 0)
jenkins hash: 23.76 ns per iteration (result: 3787970914)
fnv hash: 5.97 ns per iteration (result: 3875976788)
murmur hash (32 bit): 18.20 ns per iteration (result: 4065731677)

./snabb snabbmark hash 8
No PMU available: single core cpu affinity required
baseline: 3.18 ns per iteration (result: 0)
jenkins hash: 7.52 ns per iteration (result: 236129928)
fnv hash: 7.51 ns per iteration (result: 3645689972)
murmur hash (32 bit): 10.35 ns per iteration (result: 1669671676)

./snabb snabbmark hash 12
No PMU available: single core cpu affinity required
baseline: 3.23 ns per iteration (result: 0)
jenkins hash: 100.78 ns per iteration (result: 3787970914)
fnv hash: 10.37 ns per iteration (result: 1767269300)
murmur hash (32 bit): 105.52 ns per iteration (result: 3644920907)

./snabb snabbmark hash 16
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 102.95 ns per iteration (result: 1799890284)
fnv hash: 13.33 ns per iteration (result: 2754996980)
murmur hash (32 bit): 105.18 ns per iteration (result: 2167721464)

./snabb snabbmark hash 24
No PMU available: single core cpu affinity required
baseline: 3.26 ns per iteration (result: 0)
jenkins hash: 121.83 ns per iteration (result: 2548393908)
fnv hash: 20.30 ns per iteration (result: 3617052020)
murmur hash (32 bit): 123.42 ns per iteration (result: 3708130561)

./snabb snabbmark hash 32
No PMU available: single core cpu affinity required
baseline: 2.53 ns per iteration (result: 0)
jenkins hash: 127.32 ns per iteration (result: 333401968)
fnv hash: 29.89 ns per iteration (result: 666669044)
murmur hash (32 bit): 124.47 ns per iteration (result: 407949976)
```

Conclusion: Marginal improvement (I think).

wingo committed:
Add siphash implementation

wingo commented:
OK!  I just hacked up a SipHash implementation, thinking that it might be faster than FNV as it reads 8 bytes at a time, even though its internal state is larger.  Also of course SipHash is a proper cryptographic hash function and it would be nice if we used one, for security reasons, if we can afford to pay for it anyway.

Results:
```
$ for i in 1 2 4 6 8 12 16 24 32; do echo "./snabb snabbmark hash $i"; sudo ./snabb snabbmark hash $i; echo; done
./snabb snabbmark hash 1
No PMU available: single core cpu affinity required
baseline: 2.85 ns per iteration (result: 0)
jenkins hash: 18.31 ns per iteration (result: 236129928)
fnv hash: 4.10 ns per iteration (result: 2590564124)
sip hash: 13.50 ns per iteration (result: 1962424773)
murmur hash (32 bit): 17.92 ns per iteration (result: 1364076727)

./snabb snabbmark hash 2
No PMU available: single core cpu affinity required
baseline: 3.50 ns per iteration (result: 0)
jenkins hash: 24.02 ns per iteration (result: 2782625322)
fnv hash: 4.72 ns per iteration (result: 790051092)
sip hash: 14.71 ns per iteration (result: 1361195817)
murmur hash (32 bit): 18.26 ns per iteration (result: 821347078)

./snabb snabbmark hash 4
No PMU available: single core cpu affinity required
baseline: 3.15 ns per iteration (result: 0)
jenkins hash: 4.42 ns per iteration (result: 2355983250)
fnv hash: 4.72 ns per iteration (result: 3420582196)
sip hash: 15.19 ns per iteration (result: 4025522647)
murmur hash (32 bit): 8.14 ns per iteration (result: 593689054)

./snabb snabbmark hash 6
No PMU available: single core cpu affinity required
baseline: 3.15 ns per iteration (result: 0)
jenkins hash: 23.66 ns per iteration (result: 3787970914)
fnv hash: 5.86 ns per iteration (result: 3875976788)
sip hash: 17.42 ns per iteration (result: 2868053420)
murmur hash (32 bit): 18.62 ns per iteration (result: 4065731677)

./snabb snabbmark hash 8
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 7.50 ns per iteration (result: 236129928)
fnv hash: 7.46 ns per iteration (result: 3645689972)
sip hash: 17.37 ns per iteration (result: 970163484)
murmur hash (32 bit): 10.24 ns per iteration (result: 1669671676)

./snabb snabbmark hash 12
No PMU available: single core cpu affinity required
baseline: 3.17 ns per iteration (result: 0)
jenkins hash: 104.29 ns per iteration (result: 3787970914)
fnv hash: 10.35 ns per iteration (result: 1767269300)
sip hash: 18.13 ns per iteration (result: 3449823072)
murmur hash (32 bit): 12.69 ns per iteration (result: 3644920907)

./snabb snabbmark hash 16
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 107.66 ns per iteration (result: 1799890284)
fnv hash: 13.28 ns per iteration (result: 2754996980)
sip hash: 21.56 ns per iteration (result: 2560676800)
murmur hash (32 bit): 111.03 ns per iteration (result: 2167721464)

./snabb snabbmark hash 24
No PMU available: single core cpu affinity required
baseline: 3.21 ns per iteration (result: 0)
jenkins hash: 117.11 ns per iteration (result: 2548393908)
fnv hash: 20.40 ns per iteration (result: 3617052020)
sip hash: 24.96 ns per iteration (result: 3846390737)
murmur hash (32 bit): 120.53 ns per iteration (result: 3708130561)

./snabb snabbmark hash 32
No PMU available: single core cpu affinity required
baseline: 2.53 ns per iteration (result: 0)
jenkins hash: 128.71 ns per iteration (result: 333401968)
fnv hash: 29.87 ns per iteration (result: 666669044)
sip hash: 31.88 ns per iteration (result: 2307970020)
murmur hash (32 bit): 128.18 ns per iteration (result: 407949976)
```

Summary: FNV hash has uniform cost of something just less than 1ns/byte, which allows it to scale really nicely to small keys.  However as keys get larger, SipHash gets more interesting, and approximately has parity with FNV at 32-byte keys.  Not sure what the conclusion for ctable should be though.

wingo commented:
Reducing the number of rounds from 2,4 to 1,2 approximately halves SipHash cost.  Additionally since we are only hashing fixed-size keys, I think we can also tweak the handling of the tail word.  Looks like to me SipHash1,2 will be a winner.

wingo committed:
Allow customization of number of siphash rounds

wingo committed:
Allow specialization of SipHash to fixed-size inputs

wingo committed:
Add SipHash1,2 to snabbmark

wingo commented:
OK!  I did a couple things in this latest push.  I customized SipHash for fixed-size inputs -- avoiding adding a content-free word in the case of input sizes being a multiple of 8.  There is an option to generate standard hashes instead if the user wants.  I also allowed customization of the number of rounds.  Results:

```
$ for i in 1 2 4 6 8 12 16 24 32; do echo "./snabb snabbmark hash $i"; sudo ./snabb snabbmark hash $i; echo; done
./snabb snabbmark hash 1
No PMU available: single core cpu affinity required
baseline: 2.89 ns per iteration (result: 0)
jenkins hash: 18.22 ns per iteration (result: 236129928)
fnv hash: 4.17 ns per iteration (result: 2590564124)
sip hash c=1,d=2: 9.25 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 13.52 ns per iteration (result: 1919933255)
murmur hash (32 bit): 17.92 ns per iteration (result: 1364076727)

./snabb snabbmark hash 2
No PMU available: single core cpu affinity required
baseline: 3.46 ns per iteration (result: 0)
jenkins hash: 24.04 ns per iteration (result: 2782625322)
fnv hash: 4.83 ns per iteration (result: 790051092)
sip hash c=1,d=2: 9.22 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 14.00 ns per iteration (result: 1919933255)
murmur hash (32 bit): 18.29 ns per iteration (result: 821347078)

./snabb snabbmark hash 4
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 4.45 ns per iteration (result: 2355983250)
fnv hash: 4.76 ns per iteration (result: 3420582196)
sip hash c=1,d=2: 8.49 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 13.29 ns per iteration (result: 1919933255)
murmur hash (32 bit): 8.10 ns per iteration (result: 593689054)

./snabb snabbmark hash 6
No PMU available: single core cpu affinity required
baseline: 3.15 ns per iteration (result: 0)
jenkins hash: 23.68 ns per iteration (result: 3787970914)
fnv hash: 5.80 ns per iteration (result: 3875976788)
sip hash c=1,d=2: 8.82 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 14.05 ns per iteration (result: 1919933255)
murmur hash (32 bit): 18.19 ns per iteration (result: 4065731677)

./snabb snabbmark hash 8
No PMU available: single core cpu affinity required
baseline: 3.19 ns per iteration (result: 0)
jenkins hash: 7.57 ns per iteration (result: 236129928)
fnv hash: 7.56 ns per iteration (result: 3645689972)
sip hash c=1,d=2: 8.69 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 13.42 ns per iteration (result: 1919933255)
murmur hash (32 bit): 10.46 ns per iteration (result: 1669671676)

./snabb snabbmark hash 12
No PMU available: single core cpu affinity required
baseline: 3.16 ns per iteration (result: 0)
jenkins hash: 105.16 ns per iteration (result: 3787970914)
fnv hash: 10.35 ns per iteration (result: 1767269300)
sip hash c=1,d=2: 10.08 ns per iteration (result: 1967584618)
sip hash c=2,d=4: 17.17 ns per iteration (result: 2012466832)
murmur hash (32 bit): 109.94 ns per iteration (result: 3644920907)

./snabb snabbmark hash 16
No PMU available: single core cpu affinity required
baseline: 3.15 ns per iteration (result: 0)
jenkins hash: 110.24 ns per iteration (result: 1799890284)
fnv hash: 13.26 ns per iteration (result: 2754996980)
sip hash c=1,d=2: 9.63 ns per iteration (result: 1967584618)
sip hash c=2,d=4: 16.87 ns per iteration (result: 2012466832)
murmur hash (32 bit): 109.34 ns per iteration (result: 2167721464)

./snabb snabbmark hash 24
No PMU available: single core cpu affinity required
baseline: 3.17 ns per iteration (result: 0)
jenkins hash: 119.94 ns per iteration (result: 2548393908)
fnv hash: 20.24 ns per iteration (result: 3617052020)
sip hash c=1,d=2: 12.35 ns per iteration (result: 3253469309)
sip hash c=2,d=4: 21.07 ns per iteration (result: 1968342852)
murmur hash (32 bit): 120.02 ns per iteration (result: 3708130561)

./snabb snabbmark hash 32
No PMU available: single core cpu affinity required
baseline: 2.52 ns per iteration (result: 0)
jenkins hash: 130.64 ns per iteration (result: 333401968)
fnv hash: 30.05 ns per iteration (result: 666669044)
sip hash c=1,d=2: 14.07 ns per iteration (result: 1148382288)
sip hash c=2,d=4: 24.79 ns per iteration (result: 816391318)
murmur hash (32 bit): 124.84 ns per iteration (result: 407949976)
```

Conclusion: SipHash1,2 is fastest of these options from 12-byte keys, and is still respectable at 8-byte keys (an important lwaftr use case).  FNV beats it for smaller keys but SipHash still performs well.  I think we should look at using it in the ctable.  We should check if it does a good job on result key distribution.

wingo committed:
Refactor siphash implementation

wingo committed:
Add SSE and AVX2 parallelism to SipHash

This is a somewhat experimental commit, while I explore performance
options.

wingo commented:
Recent commits add SSE and AVX2 parallelization.  I haven't tested the AVX2 one yet as my desktop doesn't support it, but the SSE does run about 30% faster here.  Pushing to the branch so I can test on Snabb1 :)

wingo committed:
Fixups to make AVX2 work also

wingo commented:
I have some absolutely bonkers results with this one.  On Snabb1, this tests SSE and AVX2 versions.  Note that we expect that the SSE version is 2x the scalar version, and that the AVX2 baseline is 4x the scalar version.

```
[wingo@snabb1:~/snabb/src]$ sudo taskset -c 7 ./snabb snabbmark hash 8
baseline: 11.02 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 23.35 cycles, 7.30 ns per iteration (result: 236129928)
fnv hash: 22.91 cycles, 7.16 ns per iteration (result: 3645689972)
sip hash c=1,d=2: 27.03 cycles, 8.45 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 42.28 cycles, 13.21 ns per iteration (result: 1919933255)
murmur hash (32 bit): 35.15 cycles, 10.99 ns per iteration (result: 1669671676)
sip hash c=1,d=2 (sse): 48.87 cycles, 15.27 ns per iteration (result: 3140876123)
sip hash c=2,d=4 (sse): 84.89 cycles, 26.53 ns per iteration (result: 1790091635)
sip hash c=1,d=2 (avx2): 234.55 cycles, 73.30 ns per iteration (result: 3054760404)
sip hash c=2,d=4 (avx2): 274.28 cycles, 85.72 ns per iteration (result: 208526962)
```

I do have a bug where I am computing not quite the right result; something byte-order related that I will fix but I think otherwise it's fine for measuring performance.

Conclusion: SSE *might* provide a very very marginal benefit, but AVX2 isn't worth it, at least for this workload.  So weird.  /cc @lukego.

wingo committed:
fix up generic assembler output, except for avx2

wingo committed:
Use vzeroupper before leaving AVX2

wingo committed:
Add generic X86-64 backend to SipHash assembler

wingo commented:
Current noodle status: I fixed the perf problem by adding a VZEROUPPER before leaving the AVX2 code.  Then I implemented a scalar x86-64 backend to the generic code so that we can have one implementation of the algorithms with scalar, SSE (x2), and AVX2 (x4) backends.  Results:

```
[wingo@snabb1:~/snabb/src]$ sudo taskset -c 7 ./snabb snabbmark hash 8
baseline: 11.02 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 23.30 cycles, 7.28 ns per iteration (result: 236129928)
fnv hash: 22.90 cycles, 7.16 ns per iteration (result: 3645689972)
sip hash c=1,d=2: 27.03 cycles, 8.45 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 42.28 cycles, 13.21 ns per iteration (result: 1919933255)
sip hash c=1,d=2 (bis): 28.03 cycles, 8.76 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (bis): 42.20 cycles, 13.19 ns per iteration (result: 1919933255)
murmur hash (32 bit): 35.15 cycles, 10.99 ns per iteration (result: 1669671676)
sip hash c=1,d=2 (sse): 46.87 cycles, 14.65 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (sse): 83.32 cycles, 26.04 ns per iteration (result: 1919933255)
sip hash c=1,d=2 (avx2): 50.55 cycles, 15.80 ns per iteration (result: 1448874722)
sip hash c=2,d=4 (avx2): 85.99 cycles, 26.87 ns per iteration (result: 3708685873)

[wingo@snabb1:~/snabb/src]$ sudo taskset -c 7 ./snabb snabbmark hash 7
baseline: 11.03 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 69.39 cycles, 21.68 ns per iteration (result: 265108684)
fnv hash: 20.96 cycles, 6.55 ns per iteration (result: 59835964)
sip hash c=1,d=2: 30.16 cycles, 9.43 ns per iteration (result: 3466758993)
sip hash c=2,d=4: 47.62 cycles, 14.88 ns per iteration (result: 1919933255)
sip hash c=1,d=2 (bis): 30.18 cycles, 9.43 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (bis): 47.31 cycles, 14.79 ns per iteration (result: 1919933255)
murmur hash (32 bit): 31.62 cycles, 9.88 ns per iteration (result: 3574542049)
sip hash c=1,d=2 (sse): 74.18 cycles, 23.18 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (sse): 110.06 cycles, 34.39 ns per iteration (result: 1919933255)
sip hash c=1,d=2 (avx2): 90.35 cycles, 28.24 ns per iteration (result: 1448874722)
sip hash c=2,d=4 (avx2): 126.81 cycles, 39.63 ns per iteration (result: 3708685873)

[wingo@snabb1:~/snabb/src]$ sudo taskset -c 7 ./snabb snabbmark hash 12
baseline: 11.04 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 334.76 cycles, 104.62 ns per iteration (result: 3787970914)
fnv hash: 33.01 cycles, 10.32 ns per iteration (result: 1767269300)
sip hash c=1,d=2: 32.04 cycles, 10.01 ns per iteration (result: 1967584618)
sip hash c=2,d=4: 54.16 cycles, 16.92 ns per iteration (result: 2012466832)
sip hash c=1,d=2 (bis): 32.04 cycles, 10.01 ns per iteration (result: 1967584618)
sip hash c=2,d=4 (bis): 54.07 cycles, 16.90 ns per iteration (result: 2012466832)
murmur hash (32 bit): 43.46 cycles, 13.58 ns per iteration (result: 3644920907)
sip hash c=1,d=2 (sse): 81.98 cycles, 25.62 ns per iteration (result: 1967584618)
sip hash c=2,d=4 (sse): 130.29 cycles, 40.72 ns per iteration (result: 2012466832)
sip hash c=1,d=2 (avx2): 90.80 cycles, 28.38 ns per iteration (result: 1082496443)
sip hash c=2,d=4 (avx2): 138.89 cycles, 43.41 ns per iteration (result: 3519486915)

[wingo@snabb1:~/snabb/src]$ sudo taskset -c 7 ./snabb snabbmark hash 100
baseline: 18.03 cycles, 5.64 ns per iteration (result: 0)
jenkins hash: 705.29 cycles, 220.41 ns per iteration (result: 3881829434)
fnv hash: 381.31 cycles, 119.16 ns per iteration (result: 1178267444)
sip hash c=1,d=2: 111.18 cycles, 34.74 ns per iteration (result: 4209001927)
sip hash c=2,d=4: 198.17 cycles, 61.93 ns per iteration (result: 623428526)
sip hash c=1,d=2 (bis): 110.74 cycles, 34.61 ns per iteration (result: 4209001927)
sip hash c=2,d=4 (bis): 197.30 cycles, 61.66 ns per iteration (result: 623428526)
murmur hash (32 bit): 592.27 cycles, 185.09 ns per iteration (result: 202623366)
sip hash c=1,d=2 (sse): 214.74 cycles, 67.11 ns per iteration (result: 4209001927)
sip hash c=2,d=4 (sse): 394.86 cycles, 123.40 ns per iteration (result: 623428526)
sip hash c=1,d=2 (avx2): 245.73 cycles, 76.79 ns per iteration (result: 2661242387)
sip hash c=2,d=4 (avx2): 421.94 cycles, 131.86 ns per iteration (result: 3247911644)
```

The AVX2 result is still incorrect for reasons that elude me currently.  However given that the generic scalar implementation (the "`(bis)`" implementation) has the same perf as the hand-coded version, and there is also a Lua backend for that implementation, and the Lua and Scalar and SSE backends match output with the hand-coded versions, I will simplify in the next commit to remove the hand-coded versions.

The AVX2 SipHash1,2 impl is about 2.16x the speed of the scalar version for 8-byte inputs, taking 50 cycles to hash 4 inputs while the scalar version takes 27 cycles to hash 1 input.  For 100-byte inputs, the AVX2 impl takes 245 cycles to hash 4 inputs, while the scalar implementation takes 111 cycles; AVX2 is 1.81x as fast.

wingo committed:
Remove hand-coded SipHash and simplify snabbmark

wingo committed:
Fix AVX2 result extraction.

wingo commented:
The updated commit fixes the AVX2 result extraction, and reworks "snabbmark hash" to show per-item hash times instead of per-stride hash times, for better understandability.  On an E5-2620v3 these are my results:

```
$ for i in 1 2 4 6 8 12 16 24 32; do echo "./snabb snabbmark hash $i"; taskset -c 7 sudo ./snabb snabbmark hash $i; echo; done
./snabb snabbmark hash 1
baseline: 11.07 cycles, 3.46 ns per iteration (result: 0)
jenkins hash: 59.07 cycles, 18.46 ns per iteration (result: 236129928)
fnv hash: 15.02 cycles, 4.69 ns per iteration (result: 2590564124)
murmur hash (32 bit): 59.21 cycles, 18.50 ns per iteration (result: 1364076727)
sip hash c=1,d=2 (x1): 29.03 cycles, 9.07 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x2): 22.91 cycles, 7.16 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x4): 13.22 cycles, 4.13 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (x1): 43.23 cycles, 13.51 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x2): 40.93 cycles, 12.79 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x4): 21.87 cycles, 6.83 ns per iteration (result: 1919933255)

./snabb snabbmark hash 2
baseline: 10.02 cycles, 3.13 ns per iteration (result: 0)
jenkins hash: 72.85 cycles, 22.77 ns per iteration (result: 2782625322)
fnv hash: 14.02 cycles, 4.38 ns per iteration (result: 790051092)
murmur hash (32 bit): 56.87 cycles, 17.77 ns per iteration (result: 821347078)
sip hash c=1,d=2 (x1): 28.03 cycles, 8.76 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x2): 22.64 cycles, 7.07 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x4): 12.86 cycles, 4.02 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (x1): 43.37 cycles, 13.55 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x2): 40.71 cycles, 12.72 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x4): 21.65 cycles, 6.77 ns per iteration (result: 1919933255)

./snabb snabbmark hash 4
baseline: 9.04 cycles, 2.82 ns per iteration (result: 0)
jenkins hash: 13.05 cycles, 4.08 ns per iteration (result: 2355983250)
fnv hash: 14.12 cycles, 4.41 ns per iteration (result: 3420582196)
murmur hash (32 bit): 27.02 cycles, 8.44 ns per iteration (result: 593689054)
sip hash c=1,d=2 (x1): 27.03 cycles, 8.45 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x2): 22.03 cycles, 6.88 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x4): 13.34 cycles, 4.17 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (x1): 42.67 cycles, 13.33 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x2): 40.09 cycles, 12.53 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x4): 22.22 cycles, 6.94 ns per iteration (result: 1919933255)

./snabb snabbmark hash 6
baseline: 9.07 cycles, 2.84 ns per iteration (result: 0)
jenkins hash: 91.11 cycles, 28.47 ns per iteration (result: 3787970914)
fnv hash: 18.07 cycles, 5.65 ns per iteration (result: 3875976788)
murmur hash (32 bit): 63.83 cycles, 19.95 ns per iteration (result: 4065731677)
sip hash c=1,d=2 (x1): 46.71 cycles, 14.60 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x2): 23.48 cycles, 7.34 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x4): 19.16 cycles, 5.99 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (x1): 63.78 cycles, 19.93 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x2): 41.28 cycles, 12.90 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x4): 28.00 cycles, 8.75 ns per iteration (result: 1919933255)

./snabb snabbmark hash 8
baseline: 8.02 cycles, 2.51 ns per iteration (result: 0)
jenkins hash: 22.92 cycles, 7.16 ns per iteration (result: 236129928)
fnv hash: 22.02 cycles, 6.88 ns per iteration (result: 3645689972)
murmur hash (32 bit): 33.69 cycles, 10.53 ns per iteration (result: 1669671676)
sip hash c=1,d=2 (x1): 26.04 cycles, 8.14 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x2): 23.53 cycles, 7.35 ns per iteration (result: 3466758993)
sip hash c=1,d=2 (x4): 12.26 cycles, 3.83 ns per iteration (result: 3466758993)
sip hash c=2,d=4 (x1): 41.40 cycles, 12.94 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x2): 41.35 cycles, 12.92 ns per iteration (result: 1919933255)
sip hash c=2,d=4 (x4): 21.57 cycles, 6.74 ns per iteration (result: 1919933255)

./snabb snabbmark hash 12
baseline: 8.01 cycles, 2.50 ns per iteration (result: 0)
jenkins hash: 341.36 cycles, 106.68 ns per iteration (result: 3787970914)
fnv hash: 31.14 cycles, 9.73 ns per iteration (result: 1767269300)
murmur hash (32 bit): 346.23 cycles, 108.20 ns per iteration (result: 3644920907)
sip hash c=1,d=2 (x1): 52.00 cycles, 16.25 ns per iteration (result: 1967584618)
sip hash c=1,d=2 (x2): 37.54 cycles, 11.73 ns per iteration (result: 1967584618)
sip hash c=1,d=2 (x4): 21.62 cycles, 6.76 ns per iteration (result: 1967584618)
sip hash c=2,d=4 (x1): 74.27 cycles, 23.21 ns per iteration (result: 2012466832)
sip hash c=2,d=4 (x2): 61.06 cycles, 19.08 ns per iteration (result: 2012466832)
sip hash c=2,d=4 (x4): 36.84 cycles, 11.52 ns per iteration (result: 2012466832)

./snabb snabbmark hash 16
baseline: 11.04 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 356.88 cycles, 111.53 ns per iteration (result: 1799890284)
fnv hash: 42.00 cycles, 13.13 ns per iteration (result: 2754996980)
murmur hash (32 bit): 361.17 cycles, 112.87 ns per iteration (result: 2167721464)
sip hash c=1,d=2 (x1): 32.04 cycles, 10.01 ns per iteration (result: 1967584618)
sip hash c=1,d=2 (x2): 29.74 cycles, 9.30 ns per iteration (result: 1967584618)
sip hash c=1,d=2 (x4): 15.80 cycles, 4.94 ns per iteration (result: 1967584618)
sip hash c=2,d=4 (x1): 53.48 cycles, 16.71 ns per iteration (result: 2012466832)
sip hash c=2,d=4 (x2): 53.90 cycles, 16.85 ns per iteration (result: 2012466832)
sip hash c=2,d=4 (x4): 28.01 cycles, 8.75 ns per iteration (result: 2012466832)

./snabb snabbmark hash 24
baseline: 11.03 cycles, 3.45 ns per iteration (result: 0)
jenkins hash: 403.13 cycles, 125.98 ns per iteration (result: 2548393908)
fnv hash: 66.34 cycles, 20.73 ns per iteration (result: 3617052020)
murmur hash (32 bit): 380.51 cycles, 118.92 ns per iteration (result: 3708130561)
sip hash c=1,d=2 (x1): 38.35 cycles, 11.99 ns per iteration (result: 3253469309)
sip hash c=1,d=2 (x2): 35.66 cycles, 11.14 ns per iteration (result: 3253469309)
sip hash c=1,d=2 (x4): 18.92 cycles, 5.91 ns per iteration (result: 3253469309)
sip hash c=2,d=4 (x1): 65.69 cycles, 20.53 ns per iteration (result: 1968342852)
sip hash c=2,d=4 (x2): 65.75 cycles, 20.55 ns per iteration (result: 1968342852)
sip hash c=2,d=4 (x4): 34.15 cycles, 10.67 ns per iteration (result: 1968342852)

./snabb snabbmark hash 32
baseline: 11.02 cycles, 3.44 ns per iteration (result: 0)
jenkins hash: 437.52 cycles, 136.73 ns per iteration (result: 333401968)
fnv hash: 98.37 cycles, 30.74 ns per iteration (result: 666669044)
murmur hash (32 bit): 401.04 cycles, 125.33 ns per iteration (result: 407949976)
sip hash c=1,d=2 (x1): 45.78 cycles, 14.31 ns per iteration (result: 1148382288)
sip hash c=1,d=2 (x2): 42.00 cycles, 13.12 ns per iteration (result: 1148382288)
sip hash c=1,d=2 (x4): 23.66 cycles, 7.39 ns per iteration (result: 1148382288)
sip hash c=2,d=4 (x1): 79.48 cycles, 24.84 ns per iteration (result: 816391318)
sip hash c=2,d=4 (x2): 78.21 cycles, 24.44 ns per iteration (result: 816391318)
sip hash c=2,d=4 (x4): 41.67 cycles, 13.02 ns per iteration (result: 816391318)
```

I.e. basically that we get no improvement from SSE on this machine, but AVX2 is a clear win and we should integrate that into the batched ctable lookup code.

wingo commented:
Right, meant to say that SSE is the "x2" part and AVX2 is the "x4" part.

lukego commented:
Really neat! I love this workflow of implementing all the candidates on a git branch to make a direct comparison. This can also be really nice in the future if a new candidate should be considered, or the existing candidates should be re-evaluated with a new workflow, etc. Cool!

Drat that I was too late to call `VZEROUPPER`. I had that [same problem](https://software.intel.com/en-us/articles/avoiding-avx-sse-transition-penalties) not so long ago :).

Generally something I wonder about: under which circumstances should we expect 256-bit SIMD code to be twice the speed of 128-bit SIMD code?

Draft theory: when the full width of the ALU is actually used and when several instructions can be pipelined.

If the full ALU width is not being used, e.g. a memcpy routine that is not operating on the data in registers, then it seems likely the CPU can get the benefit of the wide memory paths even without using 256-bit SIMD instructions (e.g. both AVX2 and SSE memcpy routines are faster on Haswell than on Sandy Bridge... it's the width of the datapath more than the instruction set that helps.)

If you are not pipelining several instructions then there is a risk that the 256-bit instructions will have longer latencies (?) and that this will be a drag on performance when you have instructions with long dependency chains. (I need to look closer at [Agner's instruction tables](http://www.agner.org/optimize/instruction_tables.pdf) to get a feel for whether there is really a difference in the latencies or if that is just an internet rumour, I wonder if Agner's tables are available in a nicer format than PDF e.g. TXT or CSV...)

wingo committed:
siphash: Fall back to SSE if AVX2 unavailable

Clean up the SipHash implementation, separating out input stride from
size, and fall back to SSE if AVX2 isn't available.

wingo committed:
siphash: Take a table of options instead of many positional args

wingo committed:
siphash: Add a function that can hash an immediate

wingo committed:
Switch ctable over to SipHash1,2

wingo commented:
I ported the relevant parts of this PR to be a branch: https://github.com/lukego/snabb/compare/integrate-multiproc-lwaftr...Igalia:siphash-for-next.  I will submit as a PR once SnabbCo#1148 has merged.

That branch also integrates the multi-hash functionality with the ctable streamer.  Performance for the case of 4-byte keys regresses but I don't see any change for the lwaftr (which uses 8-byte keys); tested using "snabb lwaftr bench".  Of course with key sizes that are neither 4 nor 8 bytes, perf should improve substantially.
